presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -679,15 +679,17 @@ public void run(ManagedQueryExecution query)
             while (true) {
                 canQueue &= group.canQueueMore();
                 canRun &= group.canRunMore();
+                if (!canQueue && !canRun) {
+                    query.setResourceGroupQueuedOn(group.getId());
+                    query.fail(new QueryQueueFullException(id));
+                    return;
+                }
                 if (!group.parent.isPresent()) {
                     break;
                 }
                 group = group.parent.get();
             }
-            if (!canQueue && !canRun) {
-                query.fail(new QueryQueueFullException(id));
-                return;
-            }
+            query.setResourceGroupQueuedOn(this.getId());
```

Comment:
why we are setting this when the query can still be eligible to run?

Reply:
As discussed, updated such that `resourceGroupQueuedOn` will be `Optional.empty()` when eligible to run.